cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionUtils.java

```diff
@@ -353,36 +353,68 @@ public static void processAlterPartitionReassignmentsResult(AlterPartitionReassi
                                                               Set<TopicPartition> deleted,
                                                               Set<TopicPartition> dead,
                                                               Set<TopicPartition> noReassignmentToCancel) {
-    if (result != null) {
-      for (Map.Entry<TopicPartition, KafkaFuture<Void>> entry: result.values().entrySet()) {
-        TopicPartition tp = entry.getKey();
-        try {
-          entry.getValue().get(EXECUTION_TASK_FUTURE_ERROR_VERIFICATION_TIMEOUT_MS, TimeUnit.MILLISECONDS);
-          LOG.debug("Replica reassignment for {} has been accepted.", tp);
-        } catch (ExecutionException ee) {
-          if (Errors.INVALID_REPLICA_ASSIGNMENT.exception().getClass() == ee.getCause().getClass()) {
-            dead.add(tp);
-            LOG.debug("Replica reassignment failed for {} due to dead destination broker(s).", tp);
-          } else if (Errors.UNKNOWN_TOPIC_OR_PARTITION.exception().getClass() == ee.getCause().getClass()) {
-            deleted.add(tp);
-            LOG.debug("Replica reassignment failed for {} due to its topic deletion.", tp);
-          } else if (Errors.NO_REASSIGNMENT_IN_PROGRESS.exception().getClass() == ee.getCause().getClass()) {
-            // Attempt to cancel/rollback a reassignment that does not exist.
-            noReassignmentToCancel.add(tp);
-            LOG.debug("Rollback failed for {} due to lack of corresponding ongoing replica reassignment.", tp);
-          } else {
-            // Not expected to happen.
-            throw new IllegalStateException(String.format("%s encountered an unknown execution exception.", tp), ee);
-          }
-        } catch (TimeoutException | InterruptedException e) {
-          // May indicate transient (e.g. network) issues and might require a task re-execution -- i.e. handled in Executor.
-          // If this is observed frequently, we may need to bump up EXECUTION_TASK_FUTURE_ERROR_VERIFICATION_TIMEOUT_MS.
-          LOG.warn("Failed to process AlterPartitionReassignmentsResult of {}.", tp, e);
-        }
+    if (result == null) {
+      return;
+    }
+    Set<KafkaFuture<Void>> waitingToFinishKafkaFutures = new HashSet<>(result.values().values());
+    KafkaFutureResultsIterator<Void> kafkaFutureResultsIterator = new KafkaFutureResultsIterator<>(waitingToFinishKafkaFutures);
+
+    while (kafkaFutureResultsIterator.hasNext()) {
+      KafkaFutureResultsIterator.FutureResult<Void> kafkaFutureResult = kafkaFutureResultsIterator.next();
+      if (kafkaFutureResult.exceptional()) {
+        handleAlterPartitionReassignmentsException(kafkaFutureResult.exception(),
+                                                   kafkaFutureResult.finishedKafkaFuture(),
```

Comment:
To have fewer method arguments, can we pass `kafkaFutureResult` and let the retrieval of `exception()` and `finishedKafkaFuture()` happen inside the called method?